platform/util/platform-util/src/test/java/org/codice/ddf/platform/util/XMLUtilsParameterizedTests.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.commons.io.IOUtils;
+import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+import org.w3c.dom.Document;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.XMLReader;
+import org.xml.sax.helpers.DefaultHandler;
+
+@RunWith(Enclosed.class)
+public class XMLUtilsParameterizedTests {
+
+  private static final String XML_XXE_EXPANSION =
+      "<!DOCTYPE Qs [\n"
+          + "<!ENTITY Q \"Q\">\n"
+          + "<!ENTITY Q10 \"&Q;&Q;&Q;&Q;&Q;&Q;&Q;&Q;&Q;&Q;\">\n"
+          + "<!ENTITY Q100 \"&Q10;&Q10;&Q10;&Q10;&Q10;&Q10;&Q10;&Q10;&Q10;&Q10;\">\n"
+          + "<!ENTITY Q1000 \"&Q100;&Q100;&Q100;&Q100;&Q100;&Q100;&Q100;&Q100;&Q100;&Q100;\">\n"
+          + "<!ENTITY Q10000 \"&Q1000;&Q1000;&Q1000;&Q1000;&Q1000;&Q1000;&Q1000;&Q1000;&Q1000;&Q1000;\">\n"
+          + "<!ENTITY Q100000 \"&Q10000;&Q10000;&Q10000;&Q10000;&Q10000;&Q10000;&Q10000;&Q10000;&Q10000;&Q10000;\">]>\n"
+          + "<Qs>&Q100000;</Qs>\n";
+
+  private static final String XML_XXE_INJECTION =
+      "<!DOCTYPE foo [<!ENTITY bar SYSTEM \"%s\" >]>\n" + "<foo>&bar;</foo>\n";
+
+  private static final XMLUtils XML_UTILS = XMLUtils.getInstance();
+
+  @RunWith(Parameterized.class)
+  public static class TestDocumentBuilderFactoryXXE {
+
+    private String implClass;
+
+    @Parameters(name = "{index}: {0}")
+    public static Collection valuesToTestWith() {
+      return XMLUtils.DOCUMENT_BUILDER_FACTORY_IMPL_WHITELIST;
+    }
+
+    public TestDocumentBuilderFactoryXXE(String className) {
+      implClass = className;
+    }
+
+    @Test(expected = org.xml.sax.SAXParseException.class)
+    public void testDocumentBuilderLimitEntityExpansion()
+        throws IOException, SAXException, ParserConfigurationException {
+      InputStream is = new ByteArrayInputStream(XML_XXE_EXPANSION.getBytes(StandardCharsets.UTF_8));
+
+      DocumentBuilderFactory dbf =
+          XML_UTILS.getSecureDocumentBuilderFactory(implClass, XMLUtils.class.getClassLoader());
+      dbf.newDocumentBuilder().parse(is);
+    }
+
+    @Test(expected = org.xml.sax.SAXParseException.class)
+    public void testDocumentBuilderDisallowsEntityInjection()
+        throws IOException, SAXException, ParserConfigurationException {
+      URL resource = XMLUtilsTest.class.getClassLoader().getResource("xxe_injection.txt");
+      String xmlStr = String.format(XML_XXE_INJECTION, resource.toString());
+
+      InputStream is = new ByteArrayInputStream(xmlStr.getBytes(StandardCharsets.UTF_8));
+      DocumentBuilderFactory dbf =
+          XML_UTILS.getSecureDocumentBuilderFactory(implClass, XMLUtils.class.getClassLoader());
+      Document doc = dbf.newDocumentBuilder().parse(is);
+
+      // If you've gotten this far, then there's likely already an issue, but at least check that
+      // content wasn't injected
```

Comment:
If this code should never get that far, shouldn't we just rely on the expected exception to test this?

Reply:
The functionality of whether or not a given factory throws an exception seems to vary based on the implementation you use (with some other implementations just ignoring the `&bar;` element and going on their way), so that assertion is there to essentially fall back on and at least go "well, at least you accomplished the main goal and didn't inject the test file into your document"

I could probably refactor this test to be like [testXMLTransformerDisallowsEntityInjection](https://github.com/codice/ddf/pull/2468/files?diff=unified#diff-1eeafa5026a546e99739756e322b084cR135) so that it's not necessarily _relying_ on the exception, though.